File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/model/FormController.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.model;

import com.urbanairship.android.layout.event.FormEvent;
import com.urbanairship.android.layout.event.ReportingEvent;
import com.urbanairship.android.layout.property.FormBehaviorType;
import com.urbanairship.android.layout.property.ViewType;
import com.urbanairship.android.layout.reporting.FormData;
import com.urbanairship.json.JsonException;
import com.urbanairship.json.JsonMap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Controller that manages form input views.
 */
public class FormController extends BaseFormController {

    public FormController(
        @NonNull String identifier,
        @Nullable String responseType,
        @NonNull BaseModel view,
        @Nullable FormBehaviorType submitBehavior
    ) {
        super(ViewType.FORM_CONTROLLER, identifier, responseType, view, submitBehavior);
    }

    @NonNull
    public static FormController fromJson(@NonNull JsonMap json) throws JsonException {
        String identifier = identifierFromJson(json);
        BaseModel view = viewFromJson(json);
        FormBehaviorType submitBehavior = submitBehaviorFromJson(json);
        String responseType = json.opt("response_type").getString();

        return new FormController(identifier, responseType, view, submitBehavior);
    }

    @Override
    protected FormEvent.Init getInitEvent() {
        return new FormEvent.Init(getIdentifier(), isFormValid());
    }

    @Override
    protected FormEvent.DataChange getFormDataChangeEvent() {
        return new FormEvent.DataChange(new FormData.Form(getIdentifier(), getResponseType(), getFormData()), isFormValid());
    }

    @Override
    protected ReportingEvent.FormResult getFormResultEvent() {
        return new ReportingEvent.FormResult(new FormData.Form(getIdentifier(), getResponseType(), getFormData()), getFormInfo(), getAttributes());
    }

    @NonNull
    @Override
    protected String getFormType() {
        return "form";
    }

}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/model/FormController.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.model

import android.content.Context
import android.view.View
import com.urbanairship.android.layout.environment.ModelEnvironment
import com.urbanairship.android.layout.environment.SharedState
import com.urbanairship.android.layout.environment.State
import com.urbanairship.android.layout.environment.ThomasForm
import com.urbanairship.android.layout.environment.ViewEnvironment
import com.urbanairship.android.layout.info.FormControllerInfo
import com.urbanairship.android.layout.reporting.ThomasFormField

/**
 * Controller that manages form input views.
 */
internal class FormController(
    viewInfo: FormControllerInfo,
    override val view: AnyModel,
    formState: ThomasForm,
    parentState: ThomasForm?,
    pagerState: SharedState<State.Pager>?,
    environment: ModelEnvironment,
    properties: ModelProperties
) : BaseFormController<View, FormControllerInfo>(
    viewInfo = viewInfo,
    formState = formState,
    parentFormState = parentState,
    pagerState = pagerState,
    environment = environment,
    properties = properties
) {

    override fun onCreateView(
        context: Context,
        viewEnvironment: ViewEnvironment,
        itemProperties: ItemProperties?
    ) = view.createView(context, viewEnvironment, itemProperties)

    override fun buildFormData(state: State.Form) =
        ThomasFormField.Form(
            identifier = viewInfo.identifier,
            responseType = viewInfo.responseType,
            children = state.filteredFields.values.toSet(),
            filedType = ThomasFormField.FiledType.just(emptySet()))
}
